[Part 1 for #issues/27922] Add GraphQL schema API for getting repository information securely

erzhtor opened the pull request:
Part 1 of https://github.com/sourcegraph/sourcegraph/issues/27922. Blocks https://github.com/sourcegraph/sourcegraph/pull/28621.

## Description
This PR:
- Adds SQL migration to add a new index
- Updates `repositoryRedirect` GraphQL Schema API to get repository by its hashed name securely and without side-effects (like auto-cloning)

## How to test
- Run migration: `sg run migration up`
- Start Sourcegraph in DotCom mode: `sg start dotcom`
- Open [GraphQL API Console Locally](https://sourcegraph.test:3443/api/console)
- Run the following query:
```gql
  repositoryRedirect(hashedName: "any-sha256-hashed-and-hex-encoded-repository-name") {
    name
    mirrorInfo {
      cloned
    }
  }
}
```

<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->
<!-- Please notify @delivery if this PR contains changes to CI that may need to be cherry-picked on to patch release branches -->

indradhanush commented on cmd/frontend/backend/repos.go:
```diff
@@ -103,6 +103,24 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
 }
 
+// GetByHashedName retrieves the repository with the given hashed repository name
+func (s *repos) GetByHashedName(ctx context.Context, name api.HashedRepoName) (_ *types.Repo, err error) {
```
Using `(_ *types.Repo, err error)` in the return type is sending a bit of a mixed message to me as a reader. We should either use named return types or not and in this case it appears like we're trying to do both. It might be easier to just return `(*types.Repo, error)` and I'd expect rest of the code to continue working.

erzhtor replied:
I see, I was following the same convention from https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/backend/repos.go#L74. 
Also, `&err` seems to be used in the trace below. Let me know if you still think that we should still change this.

indradhanush replied:
Ah. Good catch! That makes sense and we should continue to use the named return parameters.

indradhanush commented on cmd/frontend/backend/repos.go:
```diff
@@ -103,6 +103,24 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
+	ctx, done := trace(ctx, "Repos", "GetHashedName", name, &err)
+	defer done()
+
+	repo, err := s.store.GetByHashedName(ctx, name)
+
+	if (err != nil) {
+		return nil, err
+	}
+
+	return repo, nil
+}
```
If we don't need to do any custom error handling, we could directly do:
`return s.store.GetHashedName` here.

erzhtor replied:
As mentioned https://github.com/sourcegraph/sourcegraph/pull/28387#discussion_r761096857, I think, we trace error

indradhanush replied:
Fair enough. In this case I think we can do `return repo, err` directly and won't need the `if` check.

indradhanush commented on cmd/frontend/graphqlbackend/graphqlbackend.go:
```diff
@@ -595,6 +596,19 @@ func (r *repositoryRedirect) ToRedirect() (*RedirectResolver, bool) {
 
+func (r *schemaResolver) PureRepository(ctx context.Context, args *struct {
+	HashedName *string
+}) (*RepositoryResolver, error) {
```
Is there a reason the `HashedName *string` needs to be wrapped inside a `struct`? Could we do this instead:

```
func (r *schemaResolver) PureRepository(ctx context.Context, hashedName *string)
```

erzhtor replied:
I think, it is requirement for the schema resolvers, otherwise, I'm getting:
```
fatal: expected struct or pointer to struct, got *string (hint: missing `args struct { ... }` wrapper for field arguments?)
[enterprise-frontend]   used by (*graphqlbackend.schemaResolver).PureRepository
[enterprise-frontend]   used by (*graphqlbackend.schemaResolver).Root
```

indradhanush replied:
TIL! That makes sense. Thank you!

indradhanush commented on cmd/frontend/graphqlbackend/graphqlbackend.go:
```diff
@@ -235,6 +235,7 @@ var allowedPrometheusFieldNames = map[[2]string]struct{}{
 	{"Query", "repositories"}:                   {},
 	{"Query", "repository"}:                     {},
 	{"Query", "repositoryRedirect"}:             {},
+	{"Query", "pureRepository"}:                 {},
```
Why `pureRepository`? The name feels a bit misleading, especially as we're using `hashedName` everywhere else.

erzhtor replied:
The reasoning was that `pureRepository` is exactly the same as `repository` resolver, meaning that it returns RepositoryResolver, but the difference is that it doesn't auto-clone under the hood as it happens for `repository`. 
I couldn't find a better naming for it, let me, please, know, if you have any naming suggestions?

tsenart replied:
I'd reuse `repositoryRedirect`, extending it with the necessary arguments to satisfy the new requirements.

erzhtor replied:
I like it. However, this was a suggestion from the repo-management team to create a new schema instead of updating existing `repository/repositoryRedirect` resolvers. WDYT @indradhanush, would extending `repositoryRedirect` be okay?

P.S: FYI, also, `repositoryRedirect` automatically tries to clone a repository if it doesn't exist in Sourcegraph. However, I can update logic to not clone if only hashedRepoName is passed.

indradhanush replied:
> P.S: FYI, also, repositoryRedirect automatically tries to clone a repository if it doesn't exist in Sourcegraph Cloud. However, I can update logic to not clone if only hashedRepoName is passed.

My intention was to avoid custom conditional actions on the existing API and end up with mangled code. 

I still think a separate API is a better idea here. Regarding the name, how about `hashedRepository`? What do you think  @erzhtor?

indradhanush replied:
To clarify, if you'd like to go ahead and make the changes to the existing API and think that would be simpler, I'm happy to accept that. I will leave that call on you.

tsenart replied:
> My intention was to avoid custom conditional actions on the existing API and end up with mangled code.

I don't think we would end-up with mangled code necessarily. It's easier to have one place where the same thing happens, than many, even if that thing can happen with different arguments.

indradhanush replied:
> I don't think we would end-up with mangled code necessarily. It's easier to have one place where the same thing happens, than many, even if that thing can happen with different arguments.

Hmm, fair enough. Let's go with the modifications to `repositoryRedirect` then. 

cc: @erzhtor

erzhtor replied:
Hey, @indradhanush, @tsenart, I'm updating to use the existing `repositoryRedirect` and wanted to clarify should I add this parameter to `repository` resolver as well or not?

indradhanush replied:
> I'm updating to use the existing repositoryRedirect and wanted to clarify should I add this parameter to repository resolver as well or not?

@erzhtor I don't think so, since the idea is to use `repositoryRedirect` as a proxy to find a repo by hashed name. Additionally for the requirements you won't need to use the `repository` resolver to query by hashed name I think? If you do, then by all means go for it. But overall my understanding so far is that just the changes to `repositoryRedirect` is fine. 

> P.S: I know that we can use repositoryRedirect directly, I just wanted to know if there is a convention for these resolvers to support the parameters.

Not that I'm aware of. The only resource about developing the GQL API I know about is here: https://docs.sourcegraph.com/dev/background-information/graphql_api

If you're unsure, it might be a good question for #dev-chat on Slack. :slightly_smiling_face:

indradhanush reviewed (changes requested):
Did a first pass. Could you please split this into three separate PRs to make changes incrementally:

1. PR with DB migration changes
2. PR with backend changes to the API stacked up on top of PR#1
3. PR with frontend changes stacked up on PR#2.

This would also make it easier for reviewers to move through PRs faster. :slightly_smiling_face:

erzhtor commented:
Thanks for the review, @indradhanush!

> Did a first pass. Could you please split this into three separate PRs to make changes incrementally:
> 
> 1. PR with DB migration changes
> 2. PR with backend changes to the API stacked up on top of PR#1
> 3. PR with frontend changes stacked up on PR#2.

Do you think, we can join 1st and 2nd options as a single PR, since the API won't be used until we release browser extension anyways? Otherwise, I'm just worried, that this splitting will block sequencial PRs from merging, until we release the previous ones.
-- 
A couple of additional questions, will appreciate if you could help with the answers:
1. Also, do you know the frequency of Sourcegraph Cloud releases?
2. Are those migrations run automatically when deployed? (I assume so, but want to clarify)

indradhanush commented:
> Do you think, we can join 1st and 2nd options as a single PR, since the API won't be used until we release browser extension anyways? Otherwise, I'm just worried, that this splitting will block sequencial PRs from merging, until we release the previous ones.

I thought I had added a comment about the migration in the review but can't seem to find it now. My main concern is that the alter table command is going to do a table rewrite. Our `repo` table is by far the largest table and without any benchmarks I'd prefer to err on the side of caution. I'm worried that this rewrite could take longer and incur a transaction timeout leading to manual intervention. As a result the smaller the subset of changes, the better IMO.

I understand that this does slow you down a bit, but it might be better for us to make recovery in the case of a failed DB migration as friction less as possible. 

Having said that, could you ask another pair of eyes (maybe in #dev-databases on Slack) to look and analyse the upcoming migration with this change? That would add more confidence and if they feel comfortable with it, let's go ahead and do part 1 and 2 together. Especially since it's an internal API that wont be used immediately. Also if we do go ahead with this, could you please also add a note as a headsup in that PR related to this?

> Also, do you know the frequency of Sourcegraph Cloud releases?

The frontend service on Cloud is released every 1 hour. But `repo-updater` and `gitserver` are deployed once a day.
    
> Are those migrations run automatically when deployed? (I assume so, but want to clarify)

Yes. The migrations are run automatically as part of the `sourcegraph-frontend` deployment.

tsenart commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -1116,6 +1116,18 @@ type Query {
+    an external Sourcegraph URL that may have this repository instead. Otherwise, this query returns
+    null.
+    """
+    pureRepository(
```
What's the reasoning for this query name? Also, could we instead add another parameter `hashedName` to `repositoryRedirect`?

erzhtor replied:
https://github.com/sourcegraph/sourcegraph/pull/28387#discussion_r761117914

erzhtor commented:
Hey, @indradhanush, @tsenart, thank you for the reviews! 

I've updated PR by:
- Moved browser extension changes into a separate https://github.com/sourcegraph/sourcegraph/pull/28621 PR
- Updated migration to use SQL expression index instead of generated column

I would appreciate it if you could take a look at it again?

P.S: I would also be grateful if you could help/pair on fixing Go failing tests or share some docs/hints.

tsenart commented on internal/database/repos.go:
```diff
@@ -939,6 +978,16 @@ func (s *repoStore) listSQL(ctx context.Context, opt ReposListOptions) (*sqlf.Qu
+	if len(opt.HashedNames) > 0 {
+		lowerHashedNames := make([]string, len(opt.HashedNames))
+		for i, name := range opt.HashedNames {
+			lowerHashedNames[i] = strings.ToLower(name)
```
We're lowering the hashed name instead of the name here which is a bug. If we pass the hashes to the backend, the normalization needs to happen on the client side.

tsenart commented on migrations/frontend/1528395953_add_repo_hashed_name_index.up.sql:
```diff
@@ -0,0 +1,2 @@
+-- Should run as single non-transaction block
+CREATE INDEX IF NOT EXISTS repo_hashed_name_idx ON repo USING BTREE (encode(sha256(lower(name)::bytea), 'hex'));
```
It's more efficient in terms of storage and memory usage of the DB to index on a smaller data type, so let's avoid the encoding to `hex` since it's not necessary. On the query side, we'll have to `decode('%s', 'hex')`. Additionally, since this is now a single statement, we can add back the `CONCURRENT` keyword right?

```
CREATE INDEX CONCURRENTLY IF NOT EXISTS repo_hashed_name_idx ON repo USING BTREE (sha256(lower(name)::bytea));

erzhtor replied:
> Additionally, since this is now a single statement, we can add back the CONCURRENT keyword, right?

Yep, I thought I pushed it. Fixed. Thanks! 👌 

> it's more efficient in terms of storage and memory usage of the DB to index on a smaller data type, so let's avoid the encoding to hex since it's not necessary. On the query side, we'll have to decode('%s', 'hex').

I didn't know how to construct `decode` for an array of hashed names, so used this approach. But since decided to go with the single hashed name where clause, will update to use decode instead of encoding in an index.

tsenart commented on internal/database/repos.go:
```diff
@@ -939,6 +978,16 @@ func (s *repoStore) listSQL(ctx context.Context, opt ReposListOptions) (*sqlf.Qu
 		where = append(where, sqlf.Sprintf(`lower(name::text) COLLATE "C" = ANY (%s::text[])`, pq.Array(lowerNames)))
 	}
 
+	if len(opt.HashedNames) > 0 {
```
It doesn't seem like we ever pass more than one hashed name at once so we it'd probably simplify the code to support a single one.

erzhtor replied:
Yes, for now, there is always only a single hashed name parameter. 

I can update to use a single HashedName (and it will be even simpler I think), it will be just not consistent with the rest of the RepoListOptions, which accepts an array of IDs, Names, etc. WDYT?

tsenart replied:
> which accepts an array of IDs, Names, etc. WDYT?

I think it's fine to differ here and introduce a list once it's used / needed.

tsenart reviewed (approved):
LGTM. Do you still need to make sure that we don't try to clone repos when given via hashed name?

Also, happy to pair tomorrow to run this migration in production ahead of time.

erzhtor commented:
> LGTM. Do you still need to make sure that we don't try to clone repos when given via hashed name?

Thanks for the reviews! I've made it to not clone if only hashed name passed, which I thought will work. But I've updated not clone even when passed other name/cloneURL parameters just to make it safer.

> Also, happy to pair tomorrow to run this migration in production ahead of time.

Thank you, I would be grateful for that! 👍

tsenart commented on cmd/frontend/backend/repos.go:
```diff
@@ -103,6 +103,20 @@ func (s *repos) GetByName(ctx context.Context, name api.RepoName) (_ *types.Repo
 }
 
+// GetByHashedName retrieves the repository with the given repository hashed name
+func (s *repos) GetByHashedName(ctx context.Context, name api.RepoHashedName) (_ *types.Repo, err error) {
```
Do we need this method? It seems like a pass-through. I'd simplify and call `s.store.GetByHashedName` directly in the resolver.

erzhtor replied:
I was following the rest of the codebase logic and added it here. But you're right, thank you. Updated!

indradhanush commented on cmd/frontend/graphqlbackend/graphqlbackend.go:
```diff
@@ -596,9 +597,18 @@ func (r *repositoryRedirect) ToRedirect() (*RedirectResolver, bool) {
 func (r *schemaResolver) RepositoryRedirect(ctx context.Context, args *struct {
-	Name     *string
-	CloneURL *string
+	Name       *string
+	CloneURL   *string
+	HashedName *string
```
Nit: Could we pull this struct out into a definition like:

```go
type repositoryRedirectArgs struct {
    Name       *string
    CloneURL   *string
    HashedName *string
}
```

Now that the struct has expanded a bit, it might make the code slightly cleaner.

indradhanush commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -1114,6 +1114,10 @@ type Query {
         """
         cloneURL: String
+        """
+        Query the repository by hashed name.
```
It maybe worthwhile adding that the `hashedName` is the SHA256 checksum of the absolute repo name in lower case and provide an example using `github.com/sourcegraph/sourcegraph`.

indradhanush commented on internal/database/repos.go:
```diff
@@ -233,6 +236,38 @@ func (s *repoStore) GetByName(ctx context.Context, nameOrURI api.RepoName) (_ *t
 }
 
+// GetByHashedName returns the repository with the given hashedName from the database, or an error.
+//
```
```suggestion
```

indradhanush commented on internal/database/repos.go:
```diff
@@ -233,6 +236,38 @@ func (s *repoStore) GetByName(ctx context.Context, nameOrURI api.RepoName) (_ *t
+// GetByHashedName returns the repository with the given hashedName from the database, or an error.
+//
+// RepoHashedName is the repository hashed name.
+//
```
```suggestion
//
```
```suggestion
```

indradhanush commented:
>   > Also, happy to pair tomorrow to run this migration in production ahead of time.
>
> Thank you, I would be grateful for that! +1

I might not be around to watch this (I would have loved to!), but I'll be curious to learn if we find out anything interesting from running this migration.

erzhtor commented:
@indradhanush, @tsenart, by the way, can you point out a place where we check if the user has access to see repository info or not (in graphqlbackend)? Is it done automatically somewhere in one place, or should it be part of each resolver?

P.S: I want to ensure that this hashed name access will follow the same permission checking rules, and the user will not be able to access other's private repositories.

tsenart commented:
> P.S: I want to ensure that this hashed name access will follow the same permission checking rules, and the user will not be able to access other's private repositories.

This is done in the `RepoStore` at the SQL level.

tsenart commented on internal/database/repos.go:
```diff
@@ -526,6 +559,9 @@ type ReposListOptions struct {
 	Names []string
 
+	// HashedName is a repository hashed name used to limit the results to that repository.
+	HashedName *api.RepoHashedName
```
Why are we passing a pointer? Checking for zero value of empty string is more idiomatic if that zero value is not significant.

erzhtor replied:
I'm quite new to Golang and didn't know what is the best approach here. Thank you, updated https://github.com/sourcegraph/sourcegraph/pull/28387/commits/7eb23b96e2b830b280c0541f1f3a5dc772fb0176 .
